File: packages/validator-php/src/Expr/Lexer.php
<?php

declare(strict_types=1);

namespace CRUDUI\Validator\Expr;

/**
 * CRUDUI expression lexer (EXPRESSION-GRAMMAR §1, JS Lexer parity).
 *
 * Turns a condition string into a Token[] ending in EOF. WHITESPACE is consumed
 * but never emitted. The token order, multi-char operator precedence, multi-dot
 * handling, string escapes, and number rules mirror validator-ts Lexer exactly,
 * because the shared fixture's `tokens` array is byte-compared across 4 languages.
 *
 * NO regex-split / no string-split evaluation anywhere (GRAMMAR §10): this is a
 * single forward scan. PHP identifiers are ASCII (analysis note), so byte-wise
 * scanning is exact.
 */
final class Lexer
{
    private int $position = 0;
    private readonly int $length;

    /** Build a lexer over the given expression string. */
    public function __construct(private readonly string $input)
    {
        $this->length = strlen($input);
    }

    /**
     * Tokenize the whole input. WHITESPACE tokens are dropped; an EOF token is
     * always appended last.
     *
     * @return list<Token>
     */
    public function tokenize(): array
    {
        $tokens = [];

        while (!$this->isAtEnd()) {
            $token = $this->nextToken();
            if ($token->type !== TokenType::WHITESPACE) {
                $tokens[] = $token;
            }
        }

        $tokens[] = new Token(TokenType::EOF, '', null);

        return $tokens;
    }

    private function nextToken(): Token
    {
        $start = $this->position;

        // Whitespace
        if ($this->matchWhitespace()) {
            return new Token(
                TokenType::WHITESPACE,
                substr($this->input, $start, $this->position - $start),
                null,
            );
        }

        // Multi-character operators first (order matters: longest / specific first)
        if ($this->matchString('not in') || $this->matchString('not  in')) {
            return new Token(TokenType::NOT_IN, 'not in', null);
        }
        if ($this->matchString('&&')) {
            return new Token(TokenType::AND, '&&', null);
        }
        if ($this->matchString('||')) {
            return new Token(TokenType::OR, '||', null);
        }
        if ($this->matchString('==')) {
            return new Token(TokenType::EQ, '==', null);
        }
        if ($this->matchString('!=')) {
            return new Token(TokenType::NE, '!=', null);
        }
        if ($this->matchString('>=')) {
            return new Token(TokenType::GE, '>=', null);
        }
        if ($this->matchString('<=')) {
            return new Token(TokenType::LE, '<=', null);
        }
        if ($this->matchString('>')) {
            return new Token(TokenType::GT, '>', null);
        }
        if ($this->matchString('<')) {
            return new Token(TokenType::LT, '<', null);
        }

        // NOT operator (single !), only when not part of !=
        if ($this->peek() === '!' && $this->peekNext() !== '=') {
            $this->advance();
            return new Token(TokenType::NOT, '!', null);
        }

        // Multi-dot (.., ..., etc.) vs single dot
        if ($this->peek() === '.') {
            $dotCount = 0;
            $dotStart = $this->position;
            while ($this->peek() === '.') {
                $dotCount++;
                $this->advance();
            }

            if ($dotCount > 1) {
                return new Token(
                    TokenType::DOT_DOT,
                    substr($this->input, $dotStart, $this->position - $dotStart),
                    $dotCount,
                );
            }

            return new Token(TokenType::DOT, '.', null);
        }

        if ($this->matchString('*')) {
            return new Token(TokenType::ASTERISK, '*', null);
        }
        if ($this->matchString('(')) {
            return new Token(TokenType::LPAREN, '(', null);
        }
        if ($this->matchString(')')) {
            return new Token(TokenType::RPAREN, ')', null);
        }
        if ($this->matchString('[')) {
            return new Token(TokenType::LBRACKET, '[', null);
        }
        if ($this->matchString(']')) {
            return new Token(TokenType::RBRACKET, ']', null);
        }
        if ($this->matchString(',')) {
            return new Token(TokenType::COMMA, ',', null);
        }
        if ($this->matchString('?')) {
            return new Token(TokenType::QUESTION, '?', null);
        }
        if ($this->matchString(':')) {
            return new Token(TokenType::COLON, ':', null);
        }

        // String literal
        $ch = $this->peek();
        if ($ch === "'" || $ch === '"') {
            return $this->readString();
        }

        // Number: leading digit, or leading '-' directly followed by a digit
        if ($this->isDigit($ch) || ($ch === '-' && $this->isDigit($this->peekNext()))) {
            return $this->readNumber();
        }

        // Keyword / identifier
        if ($this->isAlpha($ch)) {
            return $this->readIdentifier();
        }

        // Unknown single character
        $char = $this->advance();
        return new Token(TokenType::INVALID, $char, null);
    }

    private function readString(): Token
    {
        $quote = $this->advance();
        $value = '';

        while (!$this->isAtEnd() && $this->peek() !== $quote) {
            if ($this->peek() === '\\') {
                $this->advance();
                if (!$this->isAtEnd()) {
                    $escaped = $this->advance();
                    $value .= match ($escaped) {
                        'n'  => "\n",
                        't'  => "\t",
                        'r'  => "\r",
                        '\\' => '\\',
                        "'"  => "'",
                        '"'  => '"',
                        default => $escaped,
                    };
                }
            } else {
                $value .= $this->advance();
            }
        }

        if ($this->isAtEnd()) {
            throw new ParseError('Unterminated string literal');
        }

        $this->advance(); // closing quote

        // JS stores value = quote + content + quote (raw), literal = decoded content.
        return new Token(TokenType::STRING, $quote . $value . $quote, $value);
    }

    private function readNumber(): Token
    {
        $start = $this->position;

        if ($this->peek() === '-') {
            $this->advance();
        }

        while ($this->isDigit($this->peek())) {
            $this->advance();
        }

        if ($this->peek() === '.' && $this->isDigit($this->peekNext())) {
            $this->advance(); // consume '.'
            while ($this->isDigit($this->peek())) {
                $this->advance();
            }
        }

        // Scientific notation (JS parity; Go/Rust do not lex this, but the
        // 4-language fixture set avoids exponent cases per analysis note).
        if ($this->peek() === 'e' || $this->peek() === 'E') {
            $this->advance();
            if ($this->peek() === '+' || $this->peek() === '-') {
                $this->advance();
            }
            while ($this->isDigit($this->peek())) {
                $this->advance();
            }
        }

        $value = substr($this->input, $start, $this->position - $start);

        return new Token(TokenType::NUMBER, $value, self::parseNumberLiteral($value));
    }

    /**
     * Parse a numeric literal to int or float. JS uses parseFloat (always
     * float); for JSON byte-parity we keep an int when the source has no
     * fractional/exponent part so the fixture's integer literals stay integers.
     */
    private static function parseNumberLiteral(string $value): int|float
    {
        if (preg_match('/^-?\d+$/', $value) === 1) {
            return (int) $value;
        }

        return (float) $value;
    }

    private function readIdentifier(): Token
    {
        $start = $this->position;

        while ($this->isAlphaNumeric($this->peek())) {
            $this->advance();
        }

        $value = substr($this->input, $start, $this->position - $start);

        if ($value === 'true') {
            return new Token(TokenType::BOOLEAN, $value, true);
        }
        if ($value === 'false') {
            return new Token(TokenType::BOOLEAN, $value, false);
        }
        if ($value === 'null') {
            return new Token(TokenType::NULL, $value, null);
        }
        if ($value === 'in') {
            return new Token(TokenType::IN, $value, null);
        }
        if ($value === 'not') {
            // Lookahead for "not in" (whitespace then 'in')
            $saved = $this->position;
            $this->matchWhitespace();
            if ($this->matchString('in')) {
                return new Token(TokenType::NOT_IN, 'not in', null);
            }
            $this->position = $saved;
        }

        return new Token(TokenType::IDENTIFIER, $value, $value);
    }

    private function matchWhitespace(): bool
    {
        $matched = false;
        while (!$this->isAtEnd() && $this->isWhitespace($this->peek())) {
            $this->advance();
            $matched = true;
        }
        return $matched;
    }

    private function matchString(string $str): bool
    {
        $len = strlen($str);
        if (substr($this->input, $this->position, $len) === $str) {
            $this->position += $len;
            return true;
        }
        return false;
    }

    private function peek(): string
    {
        return $this->position < $this->length ? $this->input[$this->position] : "\0";
    }

    private function peekNext(): string
    {
        return ($this->position + 1) < $this->length ? $this->input[$this->position + 1] : "\0";
    }

    private function advance(): string
    {
        $char = $this->position < $this->length ? $this->input[$this->position] : "\0";
        $this->position++;
        return $char;
    }

    private function isAtEnd(): bool
    {
        return $this->position >= $this->length;
    }

    private function isDigit(string $c): bool
    {
        return $c >= '0' && $c <= '9';
    }

    private function isAlpha(string $c): bool
    {
        return ($c >= 'a' && $c <= 'z') || ($c >= 'A' && $c <= 'Z') || $c === '_';
    }

    private function isAlphaNumeric(string $c): bool
    {
        return $this->isAlpha($c) || $this->isDigit($c);
    }

    private function isWhitespace(string $c): bool
    {
        return $c === ' ' || $c === "\t" || $c === "\n" || $c === "\r"
            || $c === "\v" || $c === "\f";
    }
}

File: packages/validator-rust/src/expr/lexer.rs
//! CRUDUI expression lexer (EXPRESSION-GRAMMAR §1, JS/PHP `Lexer` parity).
//!
//! Turns a condition string into a `Vec<Token>` ending in EOF. WHITESPACE is
//! consumed but never emitted. The token order, multi-char operator precedence,
//! multi-dot handling, string escapes, and number rules mirror validator-js
//! `Lexer` exactly, because the shared fixture's `tokens` array is byte-compared
//! across 4 languages.
//!
//! NO regex-split / no string-split evaluation anywhere (GRAMMAR §10): a single
//! forward byte scan. Identifiers are ASCII (`[A-Za-z_][A-Za-z0-9_]*`), so
//! byte-wise scanning is exact.

use super::error::ParseError;
use super::token::{Literal, Token, TokenType};

/// Forward-scanning tokenizer over a single expression string.
pub struct Lexer {
    input: Vec<u8>,
    position: usize,
}

impl Lexer {
    /// Build a lexer over the given expression string.
    pub fn new(input: &str) -> Self {
        Lexer {
            input: input.as_bytes().to_vec(),
            position: 0,
        }
    }

    /// Tokenize the whole input. WHITESPACE tokens are dropped; an EOF token is
    /// always appended last.
    pub fn tokenize(&mut self) -> Result<Vec<Token>, ParseError> {
        let mut tokens = Vec::new();

        while !self.is_at_end() {
            let token = self.next_token()?;
            if token.token_type != TokenType::Whitespace {
                tokens.push(token);
            }
        }

        tokens.push(Token::new(TokenType::Eof, "", Literal::None));
        Ok(tokens)
    }

    fn next_token(&mut self) -> Result<Token, ParseError> {
        let start = self.position;

        // Whitespace
        if self.match_whitespace() {
            let raw = self.slice(start, self.position);
            return Ok(Token::new(TokenType::Whitespace, raw, Literal::None));
        }

        // Multi-character operators first (order matters: longest / specific first)
        if self.match_string("not in") || self.match_string("not  in") {
            return Ok(Token::new(TokenType::NotIn, "not in", Literal::None));
        }
        if self.match_string("&&") {
            return Ok(Token::new(TokenType::And, "&&", Literal::None));
        }
        if self.match_string("||") {
            return Ok(Token::new(TokenType::Or, "||", Literal::None));
        }
        if self.match_string("==") {
            return Ok(Token::new(TokenType::Eq, "==", Literal::None));
        }
        if self.match_string("!=") {
            return Ok(Token::new(TokenType::Ne, "!=", Literal::None));
        }
        if self.match_string(">=") {
            return Ok(Token::new(TokenType::Ge, ">=", Literal::None));
        }
        if self.match_string("<=") {
            return Ok(Token::new(TokenType::Le, "<=", Literal::None));
        }
        if self.match_string(">") {
            return Ok(Token::new(TokenType::Gt, ">", Literal::None));
        }
        if self.match_string("<") {
            return Ok(Token::new(TokenType::Lt, "<", Literal::None));
        }

        // NOT operator (single !), only when not part of !=
        if self.peek() == b'!' && self.peek_next() != b'=' {
            self.advance();
            return Ok(Token::new(TokenType::Not, "!", Literal::None));
        }

        // Multi-dot (.., ..., etc.) vs single dot
        if self.peek() == b'.' {
            let mut dot_count: i64 = 0;
            let dot_start = self.position;
            while self.peek() == b'.' {
                dot_count += 1;
                self.advance();
            }

            if dot_count > 1 {
                let raw = self.slice(dot_start, self.position);
                return Ok(Token::new(
                    TokenType::DotDot,
                    raw,
                    Literal::DotCount(dot_count),
                ));
            }
            return Ok(Token::new(TokenType::Dot, ".", Literal::None));
        }

        if self.match_string("*") {
            return Ok(Token::new(TokenType::Asterisk, "*", Literal::None));
        }
        if self.match_string("(") {
            return Ok(Token::new(TokenType::LParen, "(", Literal::None));
        }
        if self.match_string(")") {
            return Ok(Token::new(TokenType::RParen, ")", Literal::None));
        }
        if self.match_string("[") {
            return Ok(Token::new(TokenType::LBracket, "[", Literal::None));
        }
        if self.match_string("]") {
            return Ok(Token::new(TokenType::RBracket, "]", Literal::None));
        }
        if self.match_string(",") {
            return Ok(Token::new(TokenType::Comma, ",", Literal::None));
        }
        if self.match_string("?") {
            return Ok(Token::new(TokenType::Question, "?", Literal::None));
        }
        if self.match_string(":") {
            return Ok(Token::new(TokenType::Colon, ":", Literal::None));
        }

        // String literal
        let ch = self.peek();
        if ch == b'\'' || ch == b'"' {
            return self.read_string();
        }

        // Number: leading digit, or leading '-' directly followed by a digit
        if is_digit(ch) || (ch == b'-' && is_digit(self.peek_next())) {
            return Ok(self.read_number());
        }

        // Keyword / identifier
        if is_alpha(ch) {
            return Ok(self.read_identifier());
        }

        // Unknown single character
        let raw = self.slice(start, start + 1);
        self.advance();
        Ok(Token::new(TokenType::Invalid, raw, Literal::None))
    }

    fn read_string(&mut self) -> Result<Token, ParseError> {
        let quote = self.advance();
        let mut value = String::new();

        while !self.is_at_end() && self.peek() != quote {
            if self.peek() == b'\\' {
                self.advance();
                if !self.is_at_end() {
                    let escaped = self.advance();
                    match escaped {
                        b'n' => value.push('\n'),
                        b't' => value.push('\t'),
                        b'r' => value.push('\r'),
                        b'\\' => value.push('\\'),
                        b'\'' => value.push('\''),
                        b'"' => value.push('"'),
                        other => value.push(other as char),
                    }
                }
            } else {
                value.push(self.advance() as char);
            }
        }

        if self.is_at_end() {
            return Err(ParseError::new("Unterminated string literal"));
        }

        self.advance(); // closing quote

        // JS stores value = quote + content + quote (raw), literal = decoded.
        let q = quote as char;
        let raw = format!("{}{}{}", q, value, q);
        Ok(Token::new(TokenType::String, raw, Literal::Str(value)))
    }

    fn read_number(&mut self) -> Token {
        let start = self.position;

        if self.peek() == b'-' {
            self.advance();
        }

        while is_digit(self.peek()) {
            self.advance();
        }

        let mut is_float = false;
        if self.peek() == b'.' && is_digit(self.peek_next()) {
            is_float = true;
            self.advance(); // consume '.'
            while is_digit(self.peek()) {
                self.advance();
            }
        }

        // Scientific notation (JS parity).
        if self.peek() == b'e' || self.peek() == b'E' {
            is_float = true;
            self.advance();
            if self.peek() == b'+' || self.peek() == b'-' {
                self.advance();
            }
            while is_digit(self.peek()) {
                self.advance();
            }
        }

        let raw = self.slice(start, self.position);

        // Keep integers as integers so fixture integer literals stay integers;
        // fractional/exponent forms become floats.
        let literal = if is_float {
            Literal::Float(raw.parse::<f64>().unwrap_or(0.0))
        } else {
            match raw.parse::<i64>() {
                Ok(n) => Literal::Int(n),
                Err(_) => Literal::Float(raw.parse::<f64>().unwrap_or(0.0)),
            }
        };

        Token::new(TokenType::Number, raw, literal)
    }

    fn read_identifier(&mut self) -> Token {
        let start = self.position;

        while is_alpha_numeric(self.peek()) {
            self.advance();
        }

        let value = self.slice(start, self.position);

        match value.as_str() {
            "true" => Token::new(TokenType::Boolean, value, Literal::Bool(true)),
            "false" => Token::new(TokenType::Boolean, value, Literal::Bool(false)),
            "null" => Token::new(TokenType::Null, value, Literal::None),
            "in" => Token::new(TokenType::In, value, Literal::None),
            "not" => {
                // Lookahead for "not in" (whitespace then 'in')
                let saved = self.position;
                self.match_whitespace();
                if self.match_string("in") {
                    return Token::new(TokenType::NotIn, "not in", Literal::None);
                }
                self.position = saved;
                Token::new(TokenType::Identifier, value.clone(), Literal::Str(value))
            }
            _ => Token::new(TokenType::Identifier, value.clone(), Literal::Str(value)),
        }
    }

    fn match_whitespace(&mut self) -> bool {
        let mut matched = false;
        while !self.is_at_end() && is_whitespace(self.peek()) {
            self.advance();
            matched = true;
        }
        matched
    }

    fn match_string(&mut self, s: &str) -> bool {
        let bytes = s.as_bytes();
        let end = self.position + bytes.len();
        if end <= self.input.len() && &self.input[self.position..end] == bytes {
            self.position = end;
            return true;
        }
        false
    }

    fn slice(&self, start: usize, end: usize) -> String {
        // Tokens are ASCII-bounded; from_utf8_lossy is exact for our byte spans.
        String::from_utf8_lossy(&self.input[start..end]).into_owned()
    }

    fn peek(&self) -> u8 {
        if self.position < self.input.len() {
            self.input[self.position]
        } else {
            0
        }
    }

    fn peek_next(&self) -> u8 {
        if self.position + 1 < self.input.len() {
            self.input[self.position + 1]
        } else {
            0
        }
    }

    fn advance(&mut self) -> u8 {
        let ch = self.peek();
        self.position += 1;
        ch
    }

    fn is_at_end(&self) -> bool {
        self.position >= self.input.len()
    }
}

fn is_digit(c: u8) -> bool {
    c.is_ascii_digit()
}

fn is_alpha(c: u8) -> bool {
    c.is_ascii_alphabetic() || c == b'_'
}

fn is_alpha_numeric(c: u8) -> bool {
    is_alpha(c) || is_digit(c)
}

fn is_whitespace(c: u8) -> bool {
    matches!(c, b' ' | b'\t' | b'\n' | b'\r' | 0x0b | 0x0c)
}
